freqtrade/rpc/emc.py

```diff
@@ -0,0 +1,234 @@
+            self._sub_tasks = [
+                self._loop.create_task(self._handle_producer_connection(producer, rpc_lock))
+                for producer in self.producers
+            ]
+
+            await asyncio.gather(*self._sub_tasks)
+        except asyncio.CancelledError:
+            pass
+        finally:
+            # Stop the loop once we are done
+            self._loop.stop()
+
+    async def _handle_producer_connection(self, producer, lock):
+        """
+        Main connection loop for the consumer
+        """
+        try:
+            while True:
+                try:
+                    url, token = producer['url'], producer['ws_token']
+                    ws_url = f"{url}?token={token}"
+
+                    async with websockets.connect(ws_url) as ws:
+                        logger.info("Connection successful")
+                        channel = WebSocketChannel(ws)
+
+                        # Tell the producer we only want these topics
+                        # Should always be the first thing we send
+                        await channel.send(
+                            self.compose_consumer_request(RPCRequestType.SUBSCRIBE, self.topics)
+                        )
+
+                        # Now receive data, if none is within the time limit, ping
+                        while True:
+                            try:
+                                message = await asyncio.wait_for(
+                                    channel.recv(),
+                                    timeout=5
+                                )
+
+                                async with lock:
+                                    # Handle the data here
+                                    # We use a lock because it will call RPC methods
+                                    self.handle_producer_message(message)
+
+                            except (asyncio.TimeoutError, websockets.exceptions.ConnectionClosed):
+                                # We haven't received data yet. Check the connection and continue.
+                                try:
+                                    # ping
+                                    ping = await channel.ping()
+
+                                    await asyncio.wait_for(ping, timeout=self.ping_timeout)
+                                    logger.debug(f"Connection to {url} still alive...")
+
+                                    continue
+                                except Exception:
+                                    logger.info(
+                                        f"Ping error {url} - retrying in {self.sleep_time}s")
+                                    await asyncio.sleep(self.sleep_time)
+
+                                    break
+                            except Exception as e:
+                                logger.exception(e)
+                                continue
+                except (
+                    socket.gaierror,
```

Comment:
while i don't mind handling a gai error here (something can always go wrong) - i think we should also check upfront if the url we're about to connect is an url in a format we expect - and corresponds to a proper uri scheme.

Reply:
:+1: